File: src/ip_solver.cpp






#include <iostream>
#include <vector>
#include <string>
#include <iomanip>
#include <cmath>

using namespace std;


struct Media {
    string name;
    double adCost;
    double planCost;
    double rewardCost;
    double childCov;
    double parentCov;
    double exposure;
};

vector<Media> medias = {
    {"电视广告", 30, 9,  0,  1.2, 0.2, 70},
    {"网站广告", 15, 3,  10, 0.1, 0.7, 100},
    {"微信广告", 10, 2,  10, 0.2, 0.8, 90}
};


struct Result {
    int x1, x2, x3;
    double totalExposure;
    double adCost, planCost, rewardCost;
    double childCov, parentCov;
    int nodesExplored;
    string question;
};


Result best;
int nodesExplored = 0;

double calcExposure(int x1, int x2, int x3) {
    return medias[0].exposure * x1 + medias[1].exposure * x2 + medias[2].exposure * x3;
}

double calcAdCost(int x1, int x2, int x3, int tvCost) {
    return tvCost * x1 + medias[1].adCost * x2 + medias[2].adCost * x3;
}

double calcPlanCost(int x1, int x2, int x3) {
    return medias[0].planCost * x1 + medias[1].planCost * x2 + medias[2].planCost * x3;
}

double calcRewardCost(int x2, int x3) {
    return medias[1].rewardCost * x2 + medias[2].rewardCost * x3;
}

double calcChildCov(int x1, int x2, int x3) {
    return medias[0].childCov * x1 + medias[1].childCov * x2 + medias[2].childCov * x3;
}

double calcParentCov(int x1, int x2, int x3) {
    return medias[0].parentCov * x1 + medias[1].parentCov * x2 + medias[2].parentCov * x3;
}


Result solveQ1(int adBudget = 400, int planBudget = 100,
                int childCovTarget = 5, int parentCovTarget = 5,
                int rewardBudget = 149) {
    best.totalExposure = -1;
    best.question = "Q1";
    nodesExplored = 0;

    int x1Max = adBudget / 30;
    int x2Max = adBudget / 15;
    int x3Max = adBudget / 10;

    for (int x1 = 0; x1 <= x1Max; x1++) {
        for (int x2 = 0; x2 <= x2Max; x2++) {
            for (int x3 = 0; x3 <= x3Max; x3++) {
                nodesExplored++;

                double adCost = calcAdCost(x1, x2, x3, 30);
                double planCost = calcPlanCost(x1, x2, x3);
                double rewardCost = calcRewardCost(x2, x3);
                double childCov = calcChildCov(x1, x2, x3);
                double parentCov = calcParentCov(x1, x2, x3);

                if (adCost > adBudget) continue;
                if (planCost > planBudget) continue;
                if (rewardCost > rewardBudget) continue;
                if (childCov < childCovTarget) continue;
                if (parentCov < parentCovTarget) continue;

                double exposure = calcExposure(x1, x2, x3);
                if (exposure > best.totalExposure) {
                    best.x1 = x1; best.x2 = x2; best.x3 = x3;
                    best.totalExposure = exposure;
                    best.adCost = adCost;
                    best.planCost = planCost;
                    best.rewardCost = rewardCost;
                    best.childCov = childCov;
                    best.parentCov = parentCov;
                }
            }
        }
    }

    best.nodesExplored = nodesExplored;
    return best;
}


Result solveQ2(int adBudget = 400, int planBudget = 100, int tvCost = 25,
                int childCovTarget = 5, int parentCovTarget = 5,
                int rewardBudget = 149) {
    best.totalExposure = -1;
    best.question = "Q2";
    nodesExplored = 0;

    int x1Max = adBudget / tvCost;
    int x2Max = adBudget / 15;
    int x3Max = adBudget / 10;

    for (int x1 = 0; x1 <= x1Max; x1++) {
        for (int x2 = 0; x2 <= x2Max; x2++) {
            for (int x3 = 0; x3 <= x3Max; x3++) {
                nodesExplored++;

                double adCost = calcAdCost(x1, x2, x3, tvCost);
                double planCost = calcPlanCost(x1, x2, x3);
                double rewardCost = calcRewardCost(x2, x3);
                double childCov = calcChildCov(x1, x2, x3);
                double parentCov = calcParentCov(x1, x2, x3);

                if (adCost > adBudget) continue;
                if (planCost > planBudget) continue;
                if (rewardCost > rewardBudget) continue;
                if (childCov < childCovTarget) continue;
                if (parentCov < parentCovTarget) continue;

                double exposure = calcExposure(x1, x2, x3);
                if (exposure > best.totalExposure) {
                    best.x1 = x1; best.x2 = x2; best.x3 = x3;
                    best.totalExposure = exposure;
                    best.adCost = adCost;
                    best.planCost = planCost;
                    best.rewardCost = rewardCost;
                    best.childCov = childCov;
                    best.parentCov = parentCov;
                }
            }
        }
    }

    best.nodesExplored = nodesExplored;
    return best;
}


Result solveQ3(int adBudget = 400, int planBudget = 200,
                int childCovTarget = 5, int parentCovTarget = 5,
                int rewardBudget = 149) {
    best.totalExposure = -1;
    best.question = "Q3";
    nodesExplored = 0;

    int x1Max = adBudget / 30;
    int x2Max = adBudget / 15;
    int x3Max = adBudget / 10;

    for (int x1 = 0; x1 <= x1Max; x1++) {
        for (int x2 = 0; x2 <= x2Max; x2++) {
            for (int x3 = 0; x3 <= x3Max; x3++) {
                nodesExplored++;

                double adCost = calcAdCost(x1, x2, x3, 30);
                double planCost = calcPlanCost(x1, x2, x3);
                double rewardCost = calcRewardCost(x2, x3);
                double childCov = calcChildCov(x1, x2, x3);
                double parentCov = calcParentCov(x1, x2, x3);

                if (adCost > adBudget) continue;
                if (planCost > planBudget) continue;
                if (rewardCost > rewardBudget) continue;
                if (childCov < childCovTarget) continue;
                if (parentCov < parentCovTarget) continue;

                double exposure = calcExposure(x1, x2, x3);
                if (exposure > best.totalExposure) {
                    best.x1 = x1; best.x2 = x2; best.x3 = x3;
                    best.totalExposure = exposure;
                    best.adCost = adCost;
                    best.planCost = planCost;
                    best.rewardCost = rewardCost;
                    best.childCov = childCov;
                    best.parentCov = parentCov;
                }
            }
        }
    }

    best.nodesExplored = nodesExplored;
    return best;
}


void printResult(const Result& r) {
    cout << "\n========================================" << endl;
    cout << "  " << r.question << " 最优解" << endl;
    cout << "========================================" << endl;
    cout << "  电视广告 x1 = " << r.x1 << endl;
    cout << "  网站广告 x2 = " << r.x2 << endl;
    cout << "  微信广告 x3 = " << r.x3 << endl;
    cout << "  总曝光     = " << fixed << setprecision(0) << r.totalExposure << " 万次" << endl;
    cout << "========================================" << endl;
    cout << "  广告成本   = " << fixed << setprecision(0) << r.adCost << " 万  (上限 400万)" << endl;
    cout << "  策划成本   = " << fixed << setprecision(0) << r.planCost << " 万" << endl;
    cout << "  奖励成本   = " << fixed << setprecision(0) << r.rewardCost << " 万  (上限 149万)" << endl;
    cout << "  儿童覆盖   = " << fixed << setprecision(1) << r.childCov << " 百万 (目标 5百万)" << endl;
    cout << "  家长覆盖   = " << fixed << setprecision(1) << r.parentCov << " 百万 (目标 5百万)" << endl;
    cout << "  搜索节点   = " << r.nodesExplored << " 个" << endl;
    cout << "========================================" << endl;
}

void printHelp() {
    cout << "\n用法: adplanner [选项]" << endl;
    cout << "选项:" << endl;
    cout << "  --q1      求解第1问（策划预算100万）" << endl;
    cout << "  --q2      求解第2问（TV广告25万）" << endl;
    cout << "  --q3      求解第3问（策划预算200万）" << endl;
    cout << "  --all     求解全部三问（默认）" << endl;
    cout << "  --help    显示帮助" << endl;
    cout << "\n示例:" << endl;
    cout << "  adplanner --q1" << endl;
    cout << "  adplanner --all" << endl;
}

int main(int argc, char* argv[]) {
    cout << "\n========================================" << endl;
    cout << "  广告投放整数规划求解器" << endl;
    cout << "  算法: 分支定界枚举法" << endl;
    cout << "  问题: 优格公司早餐麦片推广案例" << endl;
    cout << "========================================" << endl;


    if (argc == 1) {
        printResult(solveQ1());
        printResult(solveQ2());
        printResult(solveQ3());
        return 0;
    }

    for (int i = 1; i < argc; i++) {
        string arg = argv[i];
        if (arg == "--q1" || arg == "-q1") {
            printResult(solveQ1());
        } else if (arg == "--q2" || arg == "-q2") {
            printResult(solveQ2());
        } else if (arg == "--q3" || arg == "-q3") {
            printResult(solveQ3());
        } else if (arg == "--all" || arg == "-a") {
            printResult(solveQ1());
            printResult(solveQ2());
            printResult(solveQ3());
        } else if (arg == "--help" || arg == "-h") {
            printHelp();
        } else {
            cout << "未知选项: " << arg << endl;
            printHelp();
        }
    }

    return 0;
}

File: solver/ip_solver.py


from typing import List, Tuple, Optional, Dict
import time


class IntOptimizer:


    def __init__(self):
        self.best_solution = None
        self.best_value = float('-inf')
        self.nodes_explored = 0

    def enumerate_solutions(
        self,
        x1_bounds: Tuple[int, int],
        x2_bounds: Tuple[int, int],
        x3_bounds: Tuple[int, int],
        constraints: List[callable],
        objective: callable,
        maximize: bool = True
    ) -> Tuple[Tuple[int, int, int], float]:


        self.best_solution = None
        self.best_value = float('-inf') if maximize else float('inf')
        self.nodes_explored = 0

        for x1 in range(x1_bounds[0], x1_bounds[1] + 1):
            for x2 in range(x2_bounds[0], x2_bounds[1] + 1):
                for x3 in range(x3_bounds[0], x3_bounds[1] + 1):
                    self.nodes_explored += 1


                    all_satisfied = all(c(x1, x2, x3) for c in constraints)
                    if not all_satisfied:
                        continue


                    value = objective(x1, x2, x3)


                    if maximize:
                        if value > self.best_value:
                            self.best_value = value
                            self.best_solution = (x1, x2, x3)
                    else:
                        if value < self.best_value:
                            self.best_value = value
                            self.best_solution = (x1, x2, x3)

        return self.best_solution, self.best_value

    def solve_q1(
        self,
        ad_budget: int = 400,
        plan_budget: int = 100,
        child_coverage: int = 5,
        parent_coverage: int = 5,
        reward_budget: int = 149
    ) -> Dict:


        def objective(x1, x2, x3):
            return 70 * x1 + 100 * x2 + 90 * x3


        constraints = [

            lambda x1, x2, x3: 30 * x1 + 15 * x2 + 10 * x3 <= ad_budget,

            lambda x1, x2, x3: 9 * x1 + 3 * x2 + 2 * x3 <= plan_budget,

            lambda x1, x2, x3: 1.2 * x1 + 0.1 * x2 + 0.2 * x3 >= child_coverage,

            lambda x1, x2, x3: 0.2 * x1 + 0.7 * x2 + 0.8 * x3 >= parent_coverage,

            lambda x1, x2, x3: 10 * x2 + 10 * x3 <= reward_budget,
        ]


        x1_max = ad_budget // 30
        x2_max = ad_budget // 15
        x3_max = ad_budget // 10

        solution, value = self.enumerate_solutions(
            (0, x1_max), (0, x2_max), (0, x3_max),
            constraints, objective, maximize=True
        )

        return self._build_result(solution, value, "Q1", plan_budget)

    def solve_q2(
        self,
        ad_budget: int = 400,
        plan_budget: int = 100,
        tv_cost: int = 25,
        child_coverage: int = 5,
        parent_coverage: int = 5,
        reward_budget: int = 149
    ) -> Dict:


        def objective(x1, x2, x3):
            return 70 * x1 + 100 * x2 + 90 * x3

        constraints = [
            lambda x1, x2, x3: tv_cost * x1 + 15 * x2 + 10 * x3 <= ad_budget,
            lambda x1, x2, x3: 9 * x1 + 3 * x2 + 2 * x3 <= plan_budget,
            lambda x1, x2, x3: 1.2 * x1 + 0.1 * x2 + 0.2 * x3 >= child_coverage,
            lambda x1, x2, x3: 0.2 * x1 + 0.7 * x2 + 0.8 * x3 >= parent_coverage,
            lambda x1, x2, x3: 10 * x2 + 10 * x3 <= reward_budget,
        ]

        x1_max = ad_budget // tv_cost
        x2_max = ad_budget // 15
        x3_max = ad_budget // 10

        solution, value = self.enumerate_solutions(
            (0, x1_max), (0, x2_max), (0, x3_max),
            constraints, objective, maximize=True
        )

        return self._build_result(solution, value, "Q2", plan_budget, tv_cost=tv_cost)

    def solve_q3(
        self,
        ad_budget: int = 400,
        plan_budget: int = 200,
        child_coverage: int = 5,
        parent_coverage: int = 5,
        reward_budget: int = 149
    ) -> Dict:


        def objective(x1, x2, x3):
            return 70 * x1 + 100 * x2 + 90 * x3

        constraints = [
            lambda x1, x2, x3: 30 * x1 + 15 * x2 + 10 * x3 <= ad_budget,
            lambda x1, x2, x3: 9 * x1 + 3 * x2 + 2 * x3 <= plan_budget,
            lambda x1, x2, x3: 1.2 * x1 + 0.1 * x2 + 0.2 * x3 >= child_coverage,
            lambda x1, x2, x3: 0.2 * x1 + 0.7 * x2 + 0.8 * x3 >= parent_coverage,
            lambda x1, x2, x3: 10 * x2 + 10 * x3 <= reward_budget,
        ]

        x1_max = ad_budget // 30
        x2_max = ad_budget // 15
        x3_max = ad_budget // 10

        solution, value = self.enumerate_solutions(
            (0, x1_max), (0, x2_max), (0, x3_max),
            constraints, objective, maximize=True
        )

        return self._build_result(solution, value, "Q3", plan_budget)

    def _build_result(
        self,
        solution: Tuple[int, int, int],
        value: float,
        question: str,
        plan_budget: int,
        tv_cost: int = 30
    ) -> Dict:

        x1, x2, x3 = solution

        ad_cost = tv_cost * x1 + 15 * x2 + 10 * x3
        plan_cost = 9 * x1 + 3 * x2 + 2 * x3
        reward_cost = 10 * x2 + 10 * x3
        child_cov = round(1.2 * x1 + 0.1 * x2 + 0.2 * x3, 1)
        parent_cov = round(0.2 * x1 + 0.7 * x2 + 0.8 * x3, 1)

        return {
            "question": question,
            "solution": {"x1": x1, "x2": x2, "x3": x3},
            "total_exposure": value,
            "ad_cost": ad_cost,
            "plan_cost": plan_cost,
            "reward_cost": reward_cost,
            "child_coverage": child_cov,
            "parent_coverage": parent_cov,
            "nodes_explored": self.nodes_explored,
        }

    def solve_all(self) -> List[Dict]:

        return [self.solve_q1(), self.solve_q2(), self.solve_q3()]


def print_result(r: Dict):

    print(f"\n{'='*40}")
    print(f"  {r['question']} 最优解")
    print(f"{'='*40}")
    print(f"  电视广告 x1 = {r['solution']['x1']}")
    print(f"  网站广告 x2 = {r['solution']['x2']}")
    print(f"  微信广告 x3 = {r['solution']['x3']}")
    print(f"  总曝光     = {r['total_exposure']} 万次")
    print(f"{'='*40}")
    print(f"  广告成本   = {r['ad_cost']} 万  (≤400)")
    print(f"  策划成本   = {r['plan_cost']} 万  (≤{r['plan_cost']})")
    print(f"  奖励成本   = {r['reward_cost']} 万  (≤149)")
    print(f"  儿童覆盖   = {r['child_coverage']} 百万 (≥5)")
    print(f"  家长覆盖   = {r['parent_coverage']} 百万 (≥5)")
    print(f"  搜索节点   = {r['nodes_explored']}")
    print(f"{'='*40}")


if __name__ == "__main__":
    solver = IntOptimizer()
    results = solver.solve_all()

    print("\n广告投放整数规划求解器")
    print("求解器: 分支定界枚举法")
    print("问题: 优格公司早餐麦片推广案例\n")

    for r in results:
        print_result(r)
